WIP: inplace upgrade: use proper MCD image

yuqi-zhang opened the pull request:
Switch to using the MCD image from the release payload of the
corresponding nodepool. This will currently only work for 4.11+
nodepools.

Will leverage https://github.com/openshift/hypershift/pull/1290
to test

enxebre commented:
/test e2e-aws-all

yuqi-zhang commented:
Seems to be a flake

/test e2e-aws-all

yuqi-zhang commented:
> failed to create infra: failed to create VPC: VpcLimitExceeded: The maximum number of VPCs has been reached.

/retest

yuqi-zhang commented:
Hmm, I tested this by hand, and I can see that when I triggered the inplace upgrade, the pod was using quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:de6026c6b51f7dfa6220fb1bd35ed0ae4309ce18487a21b31c9206f93c8d34b7 for the MCD image (which I think should be correct).

For the inplace upgrade nodepool being tested, what release image is it using? I see references to `4.11.0-0.ci-2022-04-29-080325` in the CI pod logs but I can't seem to find where the nodepools are being collected

yuqi-zhang commented:
/test e2e-aws-all

yuqi-zhang commented:
/test e2e-aws-all

although the job does have a pretty low pass rate, so maybe I shouldn't be spamming retests and taking resources from other jobs

yuqi-zhang commented:
Going to lift the WIP since this works locally, but I am not sure whether the test failures are related.

/test e2e-aws-all

davidvossel commented on hypershift-operator/controllers/nodepool/inplace.go:
```diff
@@ -230,14 +236,13 @@ func (r *NodePoolReconciler) performNodesUpgrade(ctx context.Context, hostedClus
 
-func (r *NodePoolReconciler) reconcileUpgradePod(pod *corev1.Pod, nodeName string, nodePool *hyperv1.NodePool) error {
+func (r *NodePoolReconciler) reconcileUpgradePod(pod *corev1.Pod, nodeName, mcoImage string, nodePool *hyperv1.NodePool) error {
 	// TODO (jerzhang): unhardcode some of this
```
can you remove this comment now?

yuqi-zhang commented on hypershift-operator/controllers/nodepool/inplace.go:
```diff
@@ -178,18 +179,22 @@ func (r *NodePoolReconciler) reconcileInPlaceUpgrade(ctx context.Context, hc *hy
 		}
 	}
 
+	mcoImage, hasMcoImage := releaseImage.ComponentImages()["machine-config-operator"]
+	if !hasMcoImage {
+		return fmt.Errorf("failed to get machine-config-operator image from release payload. (existing images: %v)", releaseImage.ComponentImages())
+	}
```
So the MCD image being used for the update is the one in the hosted cluster, not the management cluster. I'm not sure if I can add an explicit OCP version check to 4.11?

The current feedback loop is probably not great. The pod will start but if the MCD in the MCO image doesn't have the hypershift flag, the pod itself will fail and crashloopbackoff. I have a TODO somewhere for tracking the pod status and bubbling it to the nodepool status, but at the present moment I don't think it goes anywhere

yuqi-zhang commented:
Rebased and modified comments a bit. Is there a way I can determine if the test failures are related to this? This ran fine locally but e2e-aws-all doesn't pass here

enxebre commented:
> Rebased and modified comments a bit. Is there a way I can determine if the test failures are related to this? This ran fine locally but e2e-aws-all doesn't pass here

@yuqi-zhang thanks!
[This](https://github.com/openshift/hypershift/pull/1358) should fix e2e-aws-all. Then I'm planning to make NodePool upgrade e2e tests required, then we'll ship this.

yuqi-zhang commented:
Rebased, let's see if we pass now

/test e2e-aws-all

yuqi-zhang commented:
The image change does pass, but let's hold in favour of https://github.com/openshift/hypershift/pull/1382 first

/hold

yuqi-zhang commented:
Closing this in favour of rebasing onto https://github.com/openshift/hypershift/pull/1382
